Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [4]:
## Ваш код и ответ
print("There're %d men and %s women in dataset"%( data.sex.value_counts()["male"],data.sex.value_counts()["female"]))

There're 207 men and 96 women in dataset


 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [5]:
## Ваш код и ответ
print("%d"%(data.sex.value_counts(normalize=True)[0]*100),"% of people in the dataset are men")

68 % of people in the dataset are men


**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [6]:
## Ваш код и ответ
print("%d men and %d women have heart disease"%(data[data.target==1].sex.value_counts()[0],data[data.target==1].sex.value_counts()[1]))

93 men and 72 women have heart disease


**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [7]:
## Ваш код и ответ
print("The proportion (of total) of male patients without heart disease is %d"%(data[data.target==0].sex.value_counts()[0]/data.sex.describe()[0]*100),"%")

The proportion (of total) of male patients without heart disease is 37 %


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [8]:
## Ваш код и ответ
print("The youngest patient who has heart disease is %d years old"%(int(data[data.target==1].age.describe()["min"])))

The youngest patient who has heart disease is 29 years old


**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [9]:
## Ваш код и ответ
print("The oldest patient who doesn't have any heart disease is %d years old"%(int(data[data.target==0].age.describe()["max"])))

The oldest patient who doesn't have any heart disease is 77 years old


**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [10]:
## Ваш код и ответ
print("The youngest female patient who has heart disease is %d years old"%(int(data[(data.sex=="female")&(data.target==1)].age.describe()["min"])))

The youngest female patient who has heart disease is 34 years old


**Какой средний возраст женщин?**

In [11]:
## Ваш код и ответ
print("The average age of female patients is %d"%(round(data[data.sex=="female"].age.describe()["mean"])))

The average age of female patients is 56


**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [12]:
## Ваш код и ответ
print("The mean and standard deviations in the age of patients who have heart disease is %d and %d respectively"%(round(data[data.target==1].age.describe()["mean"]),round(data[data.target==1].age.describe()["std"])))

The mean and standard deviations in the age of patients who have heart disease is 52 and 10 respectively


** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [13]:
## Ваш код и ответ

#There are two solutions 
#The first - if we believe that all no-heart disease patients have cholesterol below average
n=data[(data.target==0)&(data.cholesterol<data.cholesterol.mean())].target.value_counts()[0] #The number of no-heart disease patients whose level of chol less than mean
overall=data[(data.target==0)].target.count() #The number of no-heart disease patients
print("Is it true that people who don’t have heart disease have less than average cholesterol?\n",n==overall,"(according to the first solution)")
      
#The second - we just count averages
if data[data.target==0].cholesterol.mean()<data.cholesterol.mean():
    print("True,according to the secon solution")
else:
    print("False (according to the second solution)")


Is it true that people who don’t have heart disease have less than average cholesterol?
 False (according to the first solution)
False (according to the second solution)


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [14]:
#Ваш код и ответ
#target - is logical
data.groupby("rest_ecg").describe(include=[np.number]).round(2).drop("target",axis=1).T.loc[(slice(None),["mean","max"]),:] #SORRY

rest_ecg                      ST-T wave abnormality  \
age                     mean                  52.91   
                        max                   71.00   
resting_blood_pressure  mean                 129.07   
                        max                  180.00   
cholesterol             mean                 237.27   
                        max                  354.00   
max_heart_rate_achieved mean                 151.96   
                        max                  194.00   
st_depression           mean                   0.88   
                        max                    5.60   
num_major_vessels       mean                   0.64   
                        max                    4.00   

rest_ecg                      left ventricular hypertrophy  normal  
age                     mean                         61.00   55.69  
                        max                          76.00   77.00  
resting_blood_pressure  mean                        140.50  134.03  
                        max                         180.00  200.00  
cholesterol             mean                        261.75  255.14  
                        max                         327.00  564.00  
max_heart_rate_achieved mean                        125.75  147.90  
                        max                         140.00  202.00  
st_depression           mean                          2.72    1.16  
                        max                           4.40    6.20  
num_major_vessels       mean                          1.00    0.82  
                        max                           3.00    4.00

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [15]:
## Ваш код и ответ
if data[(data.target==1)&(data.sex=="male")].st_depression.mean()>data[(data.target==0)&(data.sex=="female")].st_depression.mean():
    print("Men with heart disease have more ST depression than no-heart disease women")
else:
    print("Men with heart disease have less ST depression than no-heart disease women")


Men with heart disease have less ST depression than no-heart disease women


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [16]:
## Ваш код и ответ
d_rest=data.groupby("rest_ecg").cholesterol.describe().loc[:,("min","max")]
d_chest=data.groupby("chest_pain_type").cholesterol.describe().loc[:,("min","max")]
d_t=data.groupby("thalassemia").cholesterol.describe().loc[:,("min","max")]
print("Table you need\n",pd.concat([d_rest,d_chest,d_t],axis=0,keys=['rest_ecg', 'chest_pain_type',"thalassemia"]))

Table you need
                                                 min    max
rest_ecg        ST-T wave abnormality         126.0  354.0
                left ventricular hypertrophy  197.0  327.0
                normal                        149.0  564.0
chest_pain_type atypical angina               126.0  564.0
                non-anginal pain              182.0  298.0
                typical angina                131.0  409.0
thalassemia     fixed defect                  141.0  417.0
                normal                        169.0  318.0
                reversable defect             126.0  564.0


**Сколько значений может принимать каждый из категориальных признаков?**

In [17]:
## Ваш код и ответ
#target - по сути категориальный признак (???), поэтому замена значков (как в описании) для учета данного признака
mask = data[data.target==1].index
data.loc[mask,"target"]="yes"
mask = data[data.target==0].index
data.loc[mask,"target"]="no"
data.describe(include=['object']).loc["unique",:]

sex                        2
chest_pain_type            3
fasting_blood_sugar        2
rest_ecg                   3
exercise_induced_angina    2
st_slope                   2
thalassemia                3
target                     2
Name: unique, dtype: object

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [18]:
## Ваш код и ответ
col = data.describe(include=['object']).columns
fl=[]
for i in col:
    per=data.describe(include=['object'])[i].loc["freq"]/ data.describe(include=['object'])[i].loc["count"]*100 #the percentage of "top" class
    norma= 100/data.describe(include=['object'])[i].loc["unique"] #the perfect percentage for "top" class
    fl.append(per-norma)

print("Tha greatest imbalance in categorical features has %s"%(col[fl.index(max(fl))]))

Tha greatest imbalance in categorical features has fasting_blood_sugar


---